lang/terraform/terraform.py

```diff
@@ -0,0 +1,156 @@
+from talon import Context, Module, actions, settings
+
+ctx = Context()
+mod = Module()
+ctx.matches = r"""
+mode: user.terraform
+mode: user.auto_lang
+and code.language: terraform
+"""
+
+types = {
+    "string": "string",
+    "number": "number",
+    "bool": "bool",
+    "list": "list",
+    "map": "map",
+    "null": "null"
+}
+
+ctx.lists["user.code_type"] = types
+
+common_properties = {
+    "name": "name",
+    "type": "type",
+    "description": "description",
+    "default": "default",
+    "for each": "for_each",
+    "count": "count",
+    "prevent destroy": "prevent_destroy",
+    "nullable": "nullable",
+    "sensitive": "sensitive",
+    "depends on": "depends_on",
+    "provider": "provider",
+    "source": "source",
+}
+
+mod.list("terraform_common_properties", desc="Terraform Modifiers")
+ctx.lists["self.terraform_common_properties"] = common_properties
+
+module_blocks = {
+    "variable": "variable",
+    "output": "output",
+    "provider": "provider",
+    "module": "module",
+}
+
+mod.list("terraform_module_blocks", desc="Simple Terraform Blocks")
+ctx.lists["self.terraform_module_blocks"] = module_blocks
+
+
+@mod.action_class
+class Actions:
+
+    def code_terraform_module_block(text: str):
+        """Inserts a new module-related block of a given type (e.g. variable, output, provider...)"""
+
+    def code_terraform_resource(text: str):
+        """Inserts a new resource block with given name"""
+
+    def code_terraform_data_source(text: str):
+        """Inserts a new data block with given name"""
+
+
+@ctx.action_class("user")
+class UserActions:
+
+    def code_terraform_module_block(text: str):
+        actions.auto_insert(text + " \"\"")
```

Comment:
Do we want `auto_insert` here? Cc/ @rntz

Reply:
Ah, I guess I don't yet understand the difference between `insert` and `auto_insert`. I tried to mimic other files, but it's mostly cargo cult configuration, to be honest. I could not discern the difference from the wiki. Could you give me a pointer so that I don't make mistaks in the future?

Reply:
The difference between them (which you can see by the docs available when you put `actions.auto_insert` in the REPL) is that auto_insert also runs the text through `auto_format`. That in turn reads like it's intended for prose. In practice, knausj_talon hasn't implemented auto_format, so they're identical in function.

I think auto_insert shouldn't be used in any of the programming files.

Reply:
Thank you for the explanation! I'm going to replace all my use of `auto_insert` by `insert`.